### ISSUE TITLE
Bug: [no-unnecessary-type-arguments] false positives in 8.52.0

### ISSUE THREAD

**Issue by mrginglymus** [typescript-eslint/typescript-eslint#12158]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=LAKApgHgDg9gTgFwAQEsB2CxwGYEMDGYSAgnAOYDOSA3qEkgNpq4C2YAXEhQnOmQLqdcaAJ4BuUAF9QoSLERIEIqEVJkAKsrAUAPOrVUAvCXIUAfEmPVGzNqjRIA1mBExsSfacFcefJJIkQGSCQAHpQxS0KfF4oZG0AG3RkfATcFBYqBAALFCo8pDQYZDgwAEcAVxRSgBM6fBg0biRsGBhONU0VXWtWmAB%2BTm5eNDJ-CysA4IammASwADoEmDIACj6FvoBKQPCPXPyqXEiVJCw4eFA9gAEECgBaFDIi0tAZijnF5bWNgCNcOA7YJXCIASQA5AA3IilFgwaE1VDIbJYMD1RrNf5wDrkLraSw0KYhd6fJYrVZYzZtIFhCIwbJIGpgAEAGkKMDOcAucDeGNJ3wpAIWWJpQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWiIDmsALaIWqDAG1w2HImjQO0SOogBddQF8QJoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
export interface Args {
  [name: string]: any;
}

export type ArgTypes<TArgs = Args> = { [name in keyof TArgs]: string };



// typescript eslint claims this is not required
export const foo: ArgTypes<{foo?: string}> = {};

console.log(foo.foo);
// This is a type error
console.log(foo.bar);


// I've removed it here
export const bar: ArgTypes = {};

console.log(bar.foo);
// oh dear, no error
console.log(bar.bar);
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-arguments": [
      "error"
    ]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No errors

### Actual Result

Errors

### Additional Info

This regression introduced by https://github.com/typescript-eslint/typescript-eslint/pull/11918

**Comment by rdamazio** [comment]
Also seeing this.

Here's my case:
```
export enum DriveSyncState {
  DISCONNECTED = 0,
  NEEDS_SYNC = 1,
  SYNCING = 2,
  IN_SYNC = 3,
  FAILED = 4,
}
...
let state$: BehaviorSubject<DriveSyncState>;
...

// Without the type parameter, this fails to compile with "Type 'BehaviorSubject<DriveSyncState.DISCONNECTED>' is not assignable to type 'BehaviorSubject<DriveSyncState>'."
state$ = new BehaviorSubject<DriveSyncState>(DriveSyncState.DISCONNECTED);
```

Interestingly, it also started pointing out some correct cases that it didn't before, like `readonly selectedFile = model<string>('');`

**Comment by FritsvanCampen** [comment]
I'm also seeing this on the [Apollo useQuery() variables generic](https://www.apollographql.com/docs/react/api/react/useQuery#usequery-signature).

**Comment by ej-shafran** [comment]
Huh - seems like this is the result of some TypeScript quirkiness. Apparently:

```ts
type ExpectFalse<T extends false> = T;
type AreTypesEquivalent<A, B> = A extends B ? B extends A ? true : false : false;

type Test = ExpectFalse<
  AreTypesEquivalent<
    { foo?: string; },
    { [name: string]: any; }
  >
>;
```

fails to compile ([playground link](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAogHpAxsAYgQwDYGcIB4AqUEcwEAdgCZZQBmmOAfFALxT4DcAUKJFAIIAnCPnAQsMAI4BXAJYA3TOWC4+AGigAhJqz5ES5KpqgB+I8VKVqu08AFToALlr1Hz7BC7dRbMcBawECGR0d1xOKH4hEUhxaXlFMmVwiKgAb1oAewzjJyxbGTIAc3YoAF9VZIj0gG0yNABbCFz8ooBdJzQyEBLS5IZOBnYgA))...

This doesn't work even if the `any` is replaced with `unknown` (which is weird, but I guess you can technically pass `{ [name: string]: unknown; }` into a function that expects to receive `{ foo?: string }`, and vice versa...)

**Comment by Perdolique** [comment]
I have a lot of issues and false-positives after the upgrade in `.vue` files as well. Here are some examples:

```ts
interface Slots {
  readonly icon?: unknown;
}

const slots = defineSlots<Slots>(); // Slots - This is the default value for this type parameter, so it can be omitted.
```

```ts
// Map<string, ProductGroupModel> - This is the default value for this type parameter, so it can be omitted.
const groups = ref<Map<string, ProductGroupModel>>(new Map());
```

**Comment by ej-shafran** [comment]
Yeah, the Vue issues seem to be stemming from the same problem (`Map<any, any>` is assignable to `Map<string, whatever>`, and vice versa)... I guess the issue is checking `checker.isTypeAssignableTo(a, b) && checker.isTypeAssignableTo(b, a)`. If we had access to `isTypeIdenticalTo` from the internal TypeScript compiler API, that would make all of this easy...

**Comment by Perdolique** [comment]
It's not only related to Vue, actually. It's also related to some other third-party libraries that we can't control. After the upgrade to `8.57.2`, it seems like we had many false positives in different areas where such generics are correct and provide type-narrowing as expected. `8.57.1` works correctly.

**Comment by mrginglymus** [comment]
I think the comments in https://github.com/typescript-eslint/typescript-eslint/issues/12160 are very pertinent. Notwithstanding the false positives, the enhanced detection of apparently unnecessary types has great potential to harm DX by moving type errors further from where they can be understood.

**Comment by darkbasic** [comment]
I had to pin all of the following:
```
    "typescript-eslint": "8.57.1",
    "@typescript-eslint/eslint-plugin": "8.57.1",
    "@typescript-eslint/parser": "8.57.1"
```

**Comment by AQS-DTheuke** [comment]
It also causes issues when you use it with enums:

````ts
enum SomeKeys { A ,B, C}

export const someMapping = Object.freeze<
  Record<SomeKeys, string>
>({
  [SomeKeys .A]: '/doA',
  [SomeKeys .B]: '/callB',
  [SomeKeys .C]: '/c/foobar',
});
````

Removing the type information may lead to additional keys being added without being added to `someMapping`.

I'm going to disable this rule entirely.

**Comment by SukkaW** [comment]
> Huh - seems like this is the result of some TypeScript quirkiness. Apparently:
> 
> type ExpectFalse<T extends false> = T;
> type AreTypesEquivalent<A, B> = A extends B ? B extends A ? true : false : false;
> 
> type Test = ExpectFalse<
>   AreTypesEquivalent<
>     { foo?: string; },
>     { [name: string]: any; }
>   >
> >;
> fails to compile ([playground link](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAogHpAxsAYgQwDYGcIB4AqUEcwEAdgCZZQBmmOAfFALxT4DcAUKJFAIIAnCPnAQsMAI4BXAJYA3TOWC4+AGigAhJqz5ES5KpqgB+I8VKVqu08AFToALlr1Hz7BC7dRbMcBawECGR0d1xOKH4hEUhxaXlFMmVwiKgAb1oAewzjJyxbGTIAc3YoAF9VZIj0gG0yNABbCFz8ooBdJzQyEBLS5IZOBnYgA))...
> 
> This doesn't work even if the `any` is replaced with `unknown` (which is weird, but I guess you can technically pass `{ [name: string]: unknown; }` into a function that expects to receive `{ foo?: string }`, and vice versa...)

@ej-shafran The proper way to implement `AreTypesEquivalent` is:

```ts
type ExpectFalse<T extends false> = T;

type AreTypesEquivalent<X, Y> = 
  (<T>() => T extends X ? 1 : 2) extends 
  (<T>() => T extends Y ? 1 : 2) ? true : false;

type Test = ExpectFalse<
  AreTypesEquivalent<
    { foo?: string; },
    { [name: string]: any; }
  >
>;
```

This is a known way to workaround the so-called "TypeScript" quirks.

[Playground](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAogHpAxsAYgQwDYGcIB4AqUEcwEAdgCZZQBmmOAfFALxT4DcAUJ6JFAIIAnCPnAQsMAI4BXAJYA3TOWC4AGgBooATSatOUKAAoCDQwEoWTQsVKVqqqAH4oARigAuKACYLN8lSh9IxNzSzYiEn9qLSdXD28LZ2BBaWhPOmwILgMeMTZxYBZYBAhkdEzcIKERMQkZBSUyFSCDAG9aAHsOx08sZNkyAHN2KABfdRaodoBtMjQAWwhe-qGAXU80MhAR0aCGTgYuTiA)

**Comment by ej-shafran** [comment]
> > Huh - seems like this is the result of some TypeScript quirkiness. Apparently:
> > type ExpectFalse = T;
> > type AreTypesEquivalent<A, B> = A extends B ? B extends A ? true : false : false;
> > type Test = ExpectFalse<
> > AreTypesEquivalent<
> > { foo?: string; },
> > { [name: string]: any; }
> > > ;
> > > fails to compile ([playground link](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAogHpAxsAYgQwDYGcIB4AqUEcwEAdgCZZQBmmOAfFALxT4DcAUKJFAIIAnCPnAQsMAI4BXAJYA3TOWC4+AGigAhJqz5ES5KpqgB+I8VKVqu08AFToALlr1Hz7BC7dRbMcBawECGR0d1xOKH4hEUhxaXlFMmVwiKgAb1oAewzjJyxbGTIAc3YoAF9VZIj0gG0yNABbCFz8ooBdJzQyEBLS5IZOBnYgA))...
> > 
> > 
> > This doesn't work even if the `any` is replaced with `unknown` (which is weird, but I guess you can technically pass `{ [name: string]: unknown; }` into a function that expects to receive `{ foo?: string }`, and vice versa...)
> 
> [@ej-shafran](https://github.com/ej-shafran) The proper way to implement `AreTypesEquivalent` is:
> 
> type ExpectFalse<T extends false> = T;
> 
> type AreTypesEquivalent<X, Y> = 
>   (<T>() => T extends X ? 1 : 2) extends 
>   (<T>() => T extends Y ? 1 : 2) ? true : false;
> 
> type Test = ExpectFalse<
>   AreTypesEquivalent<
>     { foo?: string; },
>     { [name: string]: any; }
>   >
> >;
> 
> This is a known way to workaround the so-called "TypeScript" quirks.
> 
> [Playground](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAogHpAxsAYgQwDYGcIB4AqUEcwEAdgCZZQBmmOAfFALxT4DcAUJ6JFAIIAnCPnAQsMAI4BXAJYA3TOWC4AGgBooATSatOUKAAoCDQwEoWTQsVKVqqqAH4oARigAuKACYLN8lSh9IxNzSzYiEn9qLSdXD28LZ2BBaWhPOmwILgMeMTZxYBZYBAhkdEzcIKERMQkZBSUyFSCDAG9aAHsOx08sZNkyAHN2KABfdRaodoBtMjQAWwhe-qGAXU80MhAR0aCGTgYuTiA)

Yup, I've seen this trick before - but is there any way to do that using the compiler API?

**Comment by ncaq** [comment]
This also affects `react-hook-form`'s `useForm`, which has a default type parameter of `FieldValues = Record<string, any>`:

```typescript
// react-hook-form's signature:
declare function useForm<TFieldValues extends FieldValues = FieldValues, ...>(...): ...;
// where FieldValues = Record<string, any>

interface MyFormValues {
  name?: string;
}

// Flagged as unnecessary, but removing it loses type safety
// (falls back to Record<string, any>)
const form = useForm<MyFormValues>();
```

The root cause in `areTypesEquivalent` is that `checker.isTypeAssignableTo` is bidirectionally true between `Record<string, any>` and any object type, because `any` is assignable to/from everything. The `Any` flag guard only catches types that are themselves `any`, not types containing `any` in their structure.

**Comment by alumni** [comment]
Upgrading from `8.57.0` to `8.58.0` causes this. I think the version in the title should be `8.57.2`. Could #11918 be reverted if no other solution is planned?

**Comment by Perdolique** [comment]
8.57.2, yes. I've reverted to 8.57.1, and it works.

**Comment by OlivierZal** [comment]
Here's a working reproduction code: https://typescript-eslint.io/play/#ts=6.0.2&fileType=.ts&code=JYOwLgpgTgZghgYwgAgDIHsDmoDCUIAmE4wcANgM7IDeAUMsvnAeiGQJ7IAOcFFA7uigEAXMgpgooTPUYRmrDsgCuFaCDgBbCGIlSQMgL61aMZSARhgrZJvQIA1gB4AKgD4AFOgBu0KUQoxAAU4KCtyJwAlCAQhAicHCHZ0GGQXABoVEAcQdH4QNzdkAF4aQwBKMRcaWXwwZSgQZB8-YADkXjTaY1oiBDJQlDMLKxsyLFAPHigtQJpkACN0AnYxDGwQPEJicMpkCuCodE1gNScl9DJ5ApMAelvkSOUruS4hMConddx8IhJyChFTrmEAxCB8UKcGxgAAWp2QZFAOlosRAEmQCF%2BO1Ie1KdkcXwmmyx-0onjoDB4fEEwjEAHIqRQ6elZKp1FodMg6WyoMzuuVaN50G0EUSPNRFstVhiSbsqBU7g8AELKMByOzeaTIYBqhBwNlUFwAZQATABmAAsAFZkDDoMjUejMdtSRQTSVbPYHOLZIyaaIuYy%2BQweRptPSeXyFUKReMNuLJSsxM6-nL3RUgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWiIDmsALaIWqDJETRoHaJHBgAviF1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg2R4BDTAHN0LSpHQkATAAZpAVki8AviBVA&tokens=false

**Comment by SuperKXT** [comment]
There is another false positive in the latest version. I am unsure if I should create a new issue for it, so I am posting it here.

```ts
type TUnionType = { type: 'idle' | 'busy' } | { type: 'error', message: string };

declare const defaultVal: TUnionType | undefined;

declare const func: <T>(val: T) => T;

const val1 = func<TUnionType>(defaultVal ?? { type: 'idle' });
//    ^? TUnionType

const val2 = func(defaultVal ?? { type: 'idle' });
//    ^? { type: "error"; message: string; } | { type: string; }
```

https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=LAKApgHgDg9gTgFwAQEsB2CxwGYEMDGYSAgnAOYDOSA3qEkgNpq4C2YAXEhQnOmQLqdcaAJ4BuUAF9QoSLERIEIqEVJkAKsrAUAPOrVUAvCXIUAfEmPVGzNqjRIA1mBExsSfacFcefJJIkQUCUVDwBVNBQYNE1Qq0UtTgByFAATABswJKQAHyQkgCMAVwoRbMlcmgSVZKw4eCSAGiQ2CgpcMg4fXjQyf0DQVLB8dNw4Inxo7iQhvCL0hAA1XHTOdQiomK1KorRZ9DBUgZAhkbGJqeRsXfxOPTMACgA3FbWASksLdWPJtGmX9IARksSGuaHweg20ViYEes1w8yWKyQAH4UVUQl0Uhksv43oEAPQE%2Bj0AB6KJkIF%2B-xWACYQWD8A94YjlulUejrJjkmlMuV8aAiSSkOTQEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWiIDmsALaIWqDAG1w2HImjQO0SOogBddQF8QJoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

**Comment by bradzacher** [comment]
We have reverted the PR that introduced this new behaviour.
It will release with our next release.

https://typescript-eslint.io/users/releases

**Comment by Perdolique** [comment]
@bradzacher, I see the latest release was a few hours ago, but no mention of this issue in the changelog. Will it be in the next minor release?

**Comment by bradzacher** [comment]
It was released. The release is automated and includes all commits up until the release job is actually run.

The commit title was just prefaced with `Revert` so the automated changelog didn't include it as it didn't match the semantic commit title patterns.

**Comment by reduckted** [comment]
@bradzacher  Is it possible to amend the release notes to mention this? I just came across the false positive mentioned in https://github.com/typescript-eslint/typescript-eslint/issues/12158#issuecomment-4114963742 after I upgraded to 8.58.0. I tried to reproduce it in the playground, but because the playground is fixed on 8.58.1, I couldn't and thought I must be going crazy since the "fix" wasn't mentioned in the release notes. 😵 

P.S. I know this is a closed issue and I shouldn't really be adding to the discussion, but I didn't think it warranted opening a new issue.
